README
-=====--------------------------------------------------------
:Horde: An HTTP server designed around message-passing daemons
-=====--------------------------------------------------------

This document is a cross between a specification, a manual, and a plan of future development; these three functions will be separated out as the project matures.

Release Notes: see NEWS
License details: see COPYING

How the server works:
---------------------
The actual 'horde' process is the dispatcher, with the following functions:
	Is the initial process, holds the listen()ing socket, has a pipe() to each child produced
	When a request comes in on socket, fork() a 'net' process
	When a request comes in over a pipe, hand it off to whatever process is registered to handle it (if none are running, fork() a new one)
	When a response comes in over a pipe, send it to the appropriate recipient.

Every child process (with the exception of 'net') is a 'handler', which is registered with the dispatcher by an entry in a .horde file; typically the name should be {modulename}.horde.
The core handlers are:
	net		Calls accept() on the listening socket, does protocol with HTTP.  Calls 'ext', and 'proc' to obtain the entity body.  Makes a 'tail' call when done
	proc	Loads a resource from disk, and applies appropriate preprocessing to it.  Calls appropriate preprocessor daemons
	ext		Determines the Content-Type (MIMEtype) of a document based on its extension.  Is a daemon (rather than a libhorde function) in order to amortise the cost of reading in /etc/mime.types
	log		Writes lines to the log file (/var/log/horde).  Serialises these writes, obviating the need for locking
	stats	Tracks server stats like bytes_today, for use by eg. traffic limiters
Other handlers go in the modules/ directory; any site-specific handlers go in modules/local/.

The standard way of extending horde's capabilities is by registering processors with 'proc'; this is done by an entry in a .proc file; typically the name should be {modulename}.proc.
One preprocessor is provided as an example of how the proc system works; it is called 'pico' and emulates the dynamic rewriting functions of the "httpico" webserver, an earlier (and somewhat unsuccessful) project by the same developer.  (The main reason for implementing 'pico' was that the author had used its features throughout his own website)

Good things to note: because all the preprocessor daemons communicate only through textual IPC (using a simple RPC protocol), they can be implemented in a wide variety of languages, although it helps if you have a binding for the hmsg-related libhorde functions.

Readiness:
----------
All workers that may conceivably need to issue sub-requests must use the (ready) message to signal that they are able to take a new request.
All workers of which only one instance may exist must always be ready to accept a new request; this usually means that they never issue sub-requests and use their stdin-buffer to queue requests.
A module uses the 'only' flag to indicate in which of the above categories it resides.  At present the dispatcher assumes these categories are mutually exclusive and exhaustive.

If you need to implement functionality involving both a gatekeeper to a shared resource and the use of subrequests (eg. to gather state), you should split it into two modules, so that the gatekeeper merely becomes one of the subrequests of the application logic / driver module.

proc Rules:
-----------
The preprocessor-matching rules in .proc are expression trees based very loosely on LISP forms (except using [square brackets] instead of (parens) because the latter are already used by the horde message protocol.  In a future version the roles of these two characters may be swapped).  For instance, the rule used by 'pico' is:
	[or [= [ext]["pico"]] [and [= [subst [0][5][ctype]]["text/"]] [grep [str [body]]["<?pico"]]]]
In a more conventional notation, this would be something like
	(ext=="pico") || ((ctype[0:5]=="text/") && (grep((str)body, "<?pico")))
There are three fundamental kinds of node here: operators, constants, and variables.  Operators include things like [=] and [or], which do the obvious things with their arguments, but also more elaborate functions like [grep] and [subst], the latter of which forms a substring from its third argument based on the first two.  Most of these operators are defined to do something useful if given too many arguments; for instance [= [a][b][c]] returns (a==b)&&(a==c), while [or [a][b][c]] returns a||b||c.  In fact it should be noted that operators short-circuit where natural, particularly [and] and [or].
The constants are things like ["text/"] (a string) and [0] and [5] (integers).  There is a third data type which cannot be supplied as a constant, and that is the "BLO" or "Binary Large Object", an array of bytes which carries its length with it (instead of using NUL-termination).
Variables are provided by the calling environment (in this case 'proc', but the expression interpreter is sufficiently general to be usable elsewhere) and include such things as [ext] (the file's extension) and [ctype] (the content-type determined by 'ext').  There is also [body], which is a BLO containing the body of the resource; because [grep] only operates on strings this is cast to string type with the [str] operator (a reasonable thing to do because files of type text/* rarely contain embedded NULs).  There are also casts for [num] and [blo]; casting to/from numeric (integer) type converts like printf-family %lu.

Pitfalls:
---------
libhorde's hmsg allows embedded NULs in data, thanks to hmsg->dlen by way of new_hmsg_d().  Similarly, it allows them in htag values.  BUT... it doesn't allow them in funct or htag names, which should be purely textual.
On a related note, some preprocessors (namely, those operating only on textual data, and which will typically only be /called/ for textual data) can pretend all-the-world's-a-Cstring, and ignore the possibility of NULs - after all, who passes an image file to a preprocessor?  You may, however, feel that robustness is worth the slight extra complication involved in using hmsg->dlen.
Another way to pass binary data, rather than using new_hmsg_d(), is to use the (read) tag, and either a file on disk or a FIFO (named pipe); this will be abstracted away on the receiving side by libhorde.  Notice however that this facility means that hmsg_from_str() can block (because it's reading from a file which might be a pipe).

hmsg format:
------------
An hmsg (pronounced 'aitch-message') consists of 'funct', 'data', and a list of 'htags', each of which is a key-value pair.
'funct' and the htag names (keys) are purely textual and may not contain parens, newlines or NULs.
'data' and the htag values may contain any character; on the wire these are escaped as follows:
	NUL		-> \0
	newline	-> \n
	(		-> \[
	)		-> \]
	\		-> \\
The representation of an hmsg is as follows:
	(funct (htag-name htag-value) (htag-name htag-value) data)

Plans:
------
TODO: make the libhorde interfaces use string structs (rather than having the new_hmsg and new_hmsg_d profusion)
TODO: cull excessive idle workers
TODO: add a second form of time profiling in dispatcher, which doesn't charge a worker for the time it spends waiting for someone else (start counting when you receive a message, stop counting when you send one)
TODO: make proc pay attention to the (onfail) attribute
TODO: make 'net' handle errors by requesting (proc /500.htm), though find a way to detect loops (in case a 500 page triggers an error/bug within proc or one of its processors)
TODO: HTTP pipelining in 'net' (controlled, of course, by hst.pipeline)
TODO: Handle hostnames in GET-url (check it's ours!)
TODO: Bernstein chaining - have a launcher that binds the socket, drops privs, then execs horde.  At the moment horde could (maybe) regain privs if compromised

Sample Transcript:
------------------
This section is not entirely accurate: the following transcript, while a reasonable way of doing things, does not correspond to the current implementation (for instance, path resolution in this case actually triggers a 302; proc doesn't use a sock_un but instead reads/writes the data through disp; we now escape special chars in messages instead of hex-encoding the whole string; rather than having a hard-coded 'stats' call, we have a '(tail) call' mechanism for registering arbitrary callbacks).  The name 'disp' is used to represent the dispatcher, whose actual executable name is 'horde'.  This transcript was originally produced not for documentation but as a rough specification of "how things should work" during the early stages of development.

Example Transcript; {} denotes hex-encoding.  Internally generated message data are hex-encoded iff they contain parens or spaces or begin with a '#' (wh. is used to indicate hex-encoding).  User-input data may be hex-encoded unnecessarily.  [] denotes variables.
	disp fork() exec(path)
	disp fork() exec(stats)
	C>disp connect
	disp fork() exec(net)
	net accept()
	C>net: GET / HTTP/1.0
	net>disp: (path /)
	disp>path: (path (from net) /)
	path>disp: (path (to net) /index.html)
	disp>net: (path /index.html)
	net>disp: (proc /index.html)
	disp fork() exec(proc)
	disp>proc: (proc (from net) /index.html)
	proc reads index.html; it contains some <?pico> tags and otherwise matches no rules
	proc opens a sock_un, /tmp/.horde_[pid] (hereonin 'sock')
	proc>disp: (pico /tmp/.horde_[pid])
	disp fork() exec(pico)
	disp>pico: (pico (from proc) /tmp/.horde_[pid])
	pico>sock>proc: (waiting pico)
	proc>sock>pico: (pico {[page html data]})
	pico>disp: (uptime)
	disp>pico: (uptime {horde: 8 days, 15:37:16 | system: 73 days, 01:49:58})
	pico>sock>proc: (pico {[page html data, pico-processed]})
	pico closes sock
	proc closes sock
	proc>disp (proc (to net) {[page html data, pico-processed]})
	disp>net (proc {[page html data, pico-processed]})
	net>C: the HTTP response, 200 OK and all the data
	net>disp (stats (time [time_t]) (ip [IP]) (status 200) (bytes [bytes]) (file /index.html) /*...etc...*/ append)
	disp>stats (stats (from net) /*...etc...*/ append)
	stats>disp (stats (to net) append)
	disp>net (stats append)
	/* we assume pipelining was not requested */
	net>disp (close)
	net return()s
	stdin>disp: (shutdown)
	disp>* (shutdown)
	path return()s
	proc return()s
	stats return()s
	pico return()s
	disp>stdout: (shutdown (when now))
	disp return()s


NEWS
Release Notes for recent versions

0.0.7	2012-03-14
Improved the hmsg format; it now properly escapes special characters in both data and htag values.  Moreover, htag values may now contain embedded NULs.
rqpath now behaves correctly (it's no longer overwritten by redirects).  The response path (more like the previous semantics) is given as rspath.
Relative symlinks are now interpreted by proc as a directive to issue a 302 redirect; symlink paths containing "/./" or "/../" may cause errors or misbehaviour.

0.0.6	2012-02-09
Fixed some bugs with logging (and the (tail) call generally).

0.0.5	2012-02-09
Various useful bits of helper functionality, and extension hooks, have been added (in particular the '(tail)' call).  Some structural changes have been made, and plenty of bugs fixed.
It is now almost certain that key parts of the API will change; in particular provisions for POST and passing of binary data.

0.0.2	2011-08-28
Several improvements to the API; it's now basically sane and it may be possible to avoid further changes to existing interfaces; no promises though.
Use in production environments is still not recommended.

0.0.1	2011-08-09
Somewhat primitive alpha-quality release.  It works (it serves web pages and detects when to run the 'pico' preprocessor), but several of the interfaces are Wrong and Likely to Change.
Use of this version in production environments is not recommended; moreover, effort expended on writing extension modules is liable to be wasted if/when interfaces change.


net.c
#define _GNU_SOURCE
/*
	horde: modular http server
	Copyright (C) 2011 Edward Cree

	Licensed under GNU GPLv3+; see top of horde.c for details
	
	net: receives, parses, and handles HTTP requests
*/
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/time.h>
#include <netdb.h>
#include <arpa/inet.h>
#include <netinet/in.h>
#include <signal.h>
#include <limits.h>
#include <errno.h>

#include "bits.h"
#include "libhorde.h"

void err(unsigned int status, const char *statusmsg, const char *headers, int fd);
bool check_msgs(hstate *hst);
ssize_t sendall(int sockfd, const void *buf, size_t length, int flags);

int main(int argc, char **argv)
{
	hstate hst;
	hst_init(&hst, argc?argv[0]:"net", false);
	const char *server="horde/"HTTPD_VERSION;
	while(check_msgs(&hst));
	int newhandle;
	struct sockaddr remote;
	socklen_t addr_size=sizeof(remote);
	time_t stamp=0;
	if((newhandle=accept(3, (struct sockaddr *)&remote, &addr_size))==-1)
	{
		if(hst.debug)
		{
			fprintf(stderr, "horde: %s[%d]: ", hst.name, getpid());
			perror("accept");
		}
		hfin(EXIT_FAILURE);
		return(EXIT_FAILURE);
	}
	else
	{
		close(3); // don't need this any more
		stamp=time(NULL);
		char date[256];
		struct tm *tm = gmtime(&stamp);
		strftime(date, sizeof(date), "%F %H:%M:%S", tm);
		hmsg ac=new_hmsg("accepted", date);
		if(!ac)
		{
			if(hst.debug) fprintf(stderr, "horde: %s[%d]: 500 - allocation failure (hmsg ac): new_hmsg: %s", hst.name, getpid(), strerror(errno));
			err(500, "Internal Server Error", NULL, newhandle);
			close(newhandle);
			hfin(EXIT_FAILURE);
			return(EXIT_FAILURE);
		}
		hsend(1, ac);
		free_hmsg(ac);
		if(hst.debug) fprintf(stderr, "horde: %s[%d]: accepted at %s\n", hst.name, getpid(), date);
	}
	bool is6=false; // is this IPv6 (else v4)?
	switch(((struct sockaddr_in *)&remote)->sin_family)
	{
		case AF_INET:
			is6=false;
		break;
		case AF_INET6:
			is6=true;
		break;
		default:
			if(hst.debug) fprintf(stderr, "horde: %s[%d]: 500 - bad address family, is %u", hst.name, getpid(), ((struct sockaddr_in *)&remote)->sin_family);
			err(500, "Internal Server Error", NULL, newhandle);
			close(newhandle);
			hfin(EXIT_FAILURE);
			return(EXIT_FAILURE);
		break;
	}
	char *ip = malloc(is6?INET6_ADDRSTRLEN:INET_ADDRSTRLEN);
	if(ip) inet_ntop(is6?AF_INET6:AF_INET, is6?(const void *)&((struct sockaddr_in6 *)&remote)->sin6_addr:(const void *)&((struct sockaddr_in *)&remote)->sin_addr, ip, is6?INET6_ADDRSTRLEN:INET_ADDRSTRLEN);
	if(ip&&hst.debug) fprintf(stderr, "horde: %s[%d]: remote IP is %s\n", hst.name, getpid(), ip);
	char *buf;unsigned int bl, bi;
	init_char(&buf, &bl, &bi);
	while((bi<4) || strncmp(buf+bi-4, "\r\n\r\n", 4))
	{
		unsigned char c;
		if(recv(newhandle, &c, 1, MSG_WAITALL)!=1)
			break;
		append_char(&buf, &bl, &bi, c);
		if(bi>ULONG_MAX-3)
			break;
		if(bi==4&&buf)
		{
			if(strcmp(buf, "\r\n\r\n")==0)
			{
				buf[bi=0]=0;
			}
		}
	}
	if(!buf)
	{
		if(hst.debug) fprintf(stderr, "horde: %s[%d]: 500 - allocation failure while reading from socket\n", hst.name, getpid());
		err(500, "Internal Server Error", NULL, newhandle);
		close(newhandle);
		hfin(EXIT_FAILURE);
		return(EXIT_FAILURE);
	}
	struct timeval starttime, endtime;
	gettimeofday(&starttime, NULL);
	while(check_msgs(&hst));
	if(hst.debug) fprintf(stderr, "horde: %s[%d]: read %u bytes\n", hst.name, getpid(), bi);
	char **line=malloc(bi*sizeof(char *));
	char *next=buf, *last;
	while(*next)
	{
		unsigned int nlines=0;
		while(*next)
		{
			unsigned int ns=0;
			while(*next&&strchr("\n\r", *next)) // skip over any \n or \r, incrementing line counter if \n
			{
				if(*next=='\n' && ns++) // count a newline; if it was already nonzero,
				{
					*next++=0;
					goto eoh; // it's the end of the headers
				}
				else
					*next++=0;
			}
			last=next;
			while(!strchr("\n\r", *next)) next++; // next now points at first \r, \n or \0.
			line[nlines++]=last;
		}
		eoh:
		if(nlines==0)
		{
			if(hst.debug) fprintf(stderr, "horde: %s[%d]: 400 - empty request\n", hst.name, getpid());
			err(400, "Bad Request (Empty)", NULL, newhandle);
			close(newhandle);
			hfin(EXIT_SUCCESS);
			return(EXIT_SUCCESS);
		}
		char **n_line=realloc(line, nlines*sizeof(char *));
		if(n_line)
			line=n_line;
		// Request-Line = Method SP Request-URI SP HTTP-Version CRLF
		if(hst.debug) fprintf(stderr, "horde: %s[%d]: Request-Line: %s\n", hst.name, getpid(), line[0]);
		char *method=strtok(line[0], " ");
		char *uri=strtok(NULL, " ");
		char *ver=strtok(NULL, "");
		char *host=NULL;
		char *ref=NULL;
		char *ua=NULL;
		http_version v=get_version(ver);
		if(v==HTTP_VERSION_UNKNOWN)
		{
			err(505, "HTTP Version Not Supported", NULL, newhandle);
			if(hst.debug) fprintf(stderr, "horde: %s[%d]: 505 HTTP Version Not Supported [%s]\n", hst.name, getpid(), ver);
			close(newhandle);
			hfin(EXIT_SUCCESS); // success because it's an unrecognised version, not merely an unimplemented feature
			return(EXIT_SUCCESS);
		}
		http_method m=get_method(method);
		switch(m)
		{
			case HTTP_METHOD_GET:
				if(*uri!='/')
				{
					if(strncmp(uri, "http://", 7)!=0)
					{
						if(hst.debug) fprintf(stderr, "horde: %s[%d]: 400 Bad Request (Malformed URI) '%s'\n", hst.name, getpid(), uri);
						err(400, "Bad Request (Malformed URI)", NULL, newhandle);
						close(newhandle);
						hfin(EXIT_SUCCESS);
						return(EXIT_SUCCESS);
					}
					else
					{
						// absoluteURI
						host=uri+7;
						char *slash=strchr(host, '/');
						if(slash)
						{
							uri=slash+1;
							*slash=0;
						}
						else
							uri="/";
					}
				}
			break;
			case HTTP_METHOD_CONNECT:
				err(501, "Not Supported", NULL, newhandle);
				if(hst.debug) fprintf(stderr, "horde: %s[%d]: 501 Not Supported (%s)\n", hst.name, getpid(), method);
				close(newhandle);
				hfin(EXIT_SUCCESS); // success because it's an inapplicable method, not merely an unimplemented feature
				return(EXIT_SUCCESS);
			break;
			case HTTP_METHOD_UNKNOWN:
				err(501, "Unknown Method", NULL, newhandle);
				if(hst.debug) fprintf(stderr, "horde: %s[%d]: 501 Unknown Method (%s(\n", hst.name, getpid(), method);
				close(newhandle);
				hfin(EXIT_SUCCESS); // success because it's an unrecognised method, not merely an unimplemented feature
				return(EXIT_SUCCESS);
			break;
			default:
				err(501, "Not Implemented", NULL, newhandle);
				if(hst.debug) fprintf(stderr, "horde: %s[%d]: 501 Not Implemented (%s)\n", hst.name, getpid(), method);
				close(newhandle);
				hfin(EXIT_FAILURE);
				return(EXIT_FAILURE);
			break;
		}
		struct hdr {http_header name; const char *un_name; const char *value;} *headers=malloc(nlines*sizeof(struct hdr));
		if(!headers)
		{
			if(hst.debug) fprintf(stderr, "horde: %s[%d]: 500 - allocation failure (struct hdr *headers): malloc: %s\n", hst.name, getpid(), strerror(errno));
			err(500, "Internal Server Error", NULL, newhandle);
			close(newhandle);
			hfin(EXIT_FAILURE);
			return(EXIT_FAILURE);
		}
		unsigned int l, nhdrs=0;
		for(l=1;l<nlines;l++)
		{
			//fprintf(stderr, "> %s\n", line[l]);
			char *colon=strchr(line[l], ':');
			if(colon)
			{
				*colon++=0;
				while(isspace(*colon)) colon++;
				http_header h=get_header(line[l]);
				headers[nhdrs++]=(struct hdr){.name=h, .un_name=line[l], .value=colon};
				if(h==HTTP_HEADER_HOST)
					host=colon;
				else if(h==HTTP_HEADER_REFERER)
					ref=colon;
				else if(h==HTTP_HEADER_USER_AGENT)
					ua=colon;
			} // otherwise, we just ignore the line
		}
		while(check_msgs(&hst));
		if(hst.shutdown)
		{
			if(hst.debug) fprintf(stderr, "horde: %s[%d]: 503 - server is shutting down\n", hst.name, getpid());
			err(503, "Service Unavailable", NULL, newhandle);
			close(newhandle);
			hfin(EXIT_SUCCESS);
			return(EXIT_SUCCESS);
		}
		switch(m)
		{
			case HTTP_METHOD_GET:;
				if(hst.debug) fprintf(stderr, "horde: %s[%d]: handling GET request (%s) (%u headers)\n", hst.name, getpid(), uri, nhdrs);
				char *rpath=normalise_path(uri);
				if(hst.debug) fprintf(stderr, "horde: %s[%d]: sending request to proc for %s\n", hst.name, getpid(), rpath);
				hmsg p=new_hmsg("proc", rpath);
				if(!p)
				{
					if(hst.debug) fprintf(stderr, "horde: %s[%d]: 500 - allocation failure (hmsg p): new_hmsg: %s\n", hst.name, getpid(), strerror(errno));
					err(500, "Internal Server Error", NULL, newhandle);
					close(newhandle);
					hfin(EXIT_FAILURE);
					return(EXIT_FAILURE);
				}
				unsigned int hdr;
				for(hdr=0;hdr<nhdrs;hdr++)
				{
					char *hline=malloc(strlen(headers[hdr].un_name)+strlen(headers[hdr].value)+3);
					if(hline)
					{
						sprintf(hline, "%s: %s", headers[hdr].un_name, headers[hdr].value);
						add_htag(p, "rqheader", hline);
						free(hline);
					}
					else
					{
						if(hst.debug) fprintf(stderr, "horde: %s[%d]: ign - allocation failure (char *hline): malloc: %s\n", hst.name, getpid(), strerror(errno));
					}
				}
				hsend(1, p);
				free_hmsg(p);
				hmsg h;
				do
				{
					if(hst.shutdown)
					{
						if(hst.debug) fprintf(stderr, "horde: %s[%d]: 503 - server is shutting down\n", hst.name, getpid());
						err(503, "Service Unavailable", NULL, newhandle);
						close(newhandle);
						hfin(EXIT_SUCCESS);
						return(EXIT_SUCCESS);
					}
					char *fromproc=getl(STDIN_FILENO);
					if(!(fromproc&&*fromproc))
					{
						if(hst.debug) fprintf(stderr, "horde: %s[%d]: 500 - failed to read response (getl): %s\n", hst.name, getpid(), strerror(errno));
						err(500, "Internal Server Error", NULL, newhandle);
						close(newhandle);
						hfin(EXIT_FAILURE);
						return(EXIT_FAILURE);
					}
					h=hmsg_from_str(fromproc, true);
					if(h)
						free(fromproc);
					else
					{
						if(hst.debug) fprintf(stderr, "horde: %s[%d]: 500 - couldn't understand the response from proc: %s\n", hst.name, getpid(), fromproc);
						err(500, "Internal Server Error", NULL, newhandle);
						close(newhandle);
						hfin(EXIT_FAILURE);
						return(EXIT_FAILURE);
					}
				}
				while(hmsg_state(h, &hst));
				const char *from=gettag(h, "from");
				if(strcmp(h->funct, "err")==0)
				{
					if(hst.debug)
					{
						fprintf(stderr, "horde: %s[%d]: 500 - proc failed: %s\n", hst.name, getpid(), h->funct);
						unsigned int i;
						for(i=0;i<h->nparms;i++)
						{
							fprintf(stderr, "horde: %s[%d]:\t(%s|%s)\n", hst.name, getpid(), h->p_tag[i], h->p_value[i]);
							if(strcmp(h->p_tag[i], "errno")==0)
							{
								fprintf(stderr, "horde: %s[%d]:\t\t%s\n", hst.name, getpid(), strerror(hgetlong(h->p_value[i])));
							}
						}
						fprintf(stderr, "horde: %s[%d]:\t%s\n", hst.name, getpid(), h->data);
					}
					err(500, "Internal Server Error", NULL, newhandle);
					close(newhandle);
					hfin(EXIT_FAILURE);
					return(EXIT_FAILURE);
				}
				else if(strcmp(h->funct, "proc")!=0)
				{
					if(hst.debug) fprintf(stderr, "horde: %s[%d]: unrecognised funct '%s'\n", hst.name, getpid(), h->funct);
					hmsg eh=new_hmsg("err", NULL);
					if(eh)
					{
						add_htag(eh, "what", "unrecognised-funct");
						if(from) add_htag(eh, "to", from);
						hsend(1, eh);
						free_hmsg(eh);
					}
					free_hmsg(h);
					close(newhandle);
					hfin(EXIT_FAILURE);
					return(EXIT_FAILURE);
				}
				unsigned short status=200;
				const char *statusmsg=gettag(h, "statusmsg");
				char *fserver; unsigned int fsl, fsi;
				init_char(&fserver, &fsl, &fsi);
				append_str(&fserver, &fsl, &fsi, server);
				unsigned int i;
				for(i=0;i<h->nparms;i++)
				{
					if(hst.debug) fprintf(stderr, "horde: %s[%d]:\t(%s|%s)\n", hst.name, getpid(), h->p_tag[i], h->p_value[i]);
					if(strcmp(h->p_tag[i], "status")==0)
					{
						unsigned short ns=hgetshort(h->p_value[i]);
						if((ns<600)&&(ns>99)) status=ns;
					}
					else if(strcmp(h->p_tag[i], "server")==0)
					{
						append_char(&fserver, &fsl, &fsi, ' ');
						append_str(&fserver, &fsl, &fsi, h->p_value[i]);
					}
				}
				if(!statusmsg)
					statusmsg=http_statusmsg(status);
				if(!statusmsg)
					statusmsg="???";
				char date[256];
				time_t timer = time(NULL);
				struct tm *tm = gmtime(&timer);
				size_t datelen = strftime(date, sizeof(date), "%F %H:%M:%S", tm);
				char *head=malloc(9+TL_SHORT+strlen(statusmsg)+1+6+datelen+1+8+strlen(fserver)+1+16+TL_SIZET+1+1);
				if(!head)
				{
					if(hst.debug) fprintf(stderr, "horde: %s[%d]: 500 - allocation failure (char *head): malloc: %s\n", hst.name, getpid(), strerror(errno));
					err(500, "Internal Server Error", NULL, newhandle);
					close(newhandle);
					hfin(EXIT_FAILURE);
					return(EXIT_FAILURE);
				}
				sprintf(head, "HTTP/1.1 %hu %s\nDate: %s\nServer: %s\nContent-Length: %zu\n", status, statusmsg, date, fserver, h->data?h->dlen:0);
				free(fserver);
				ssize_t n=sendall(newhandle, head, strlen(head), 0);
				if(n)
				{
					if(hst.debug) fprintf(stderr, "horde: %s[%d]: 499 - sendall(head) failed, %zd\n", hst.name, getpid(), n);
					close(newhandle);
					hfin(EXIT_FAILURE);
					return(EXIT_FAILURE);
				}
				free(head);
				for(i=0;i<h->nparms;i++)
				{
					if(strcmp(h->p_tag[i], "header")==0)
					{
						n=sendall(newhandle, h->p_value[i], strlen(h->p_value[i]), 0);
						if(n)
						{
							if(hst.debug) fprintf(stderr, "horde: %s[%d]: 499 - sendall(extra header) failed, %zd\n", hst.name, getpid(), n);
							close(newhandle);
							hfin(EXIT_FAILURE);
							return(EXIT_FAILURE);
						}
						if((n=sendall(newhandle, "\n", 1, 0)))
						{
							if(hst.debug) fprintf(stderr, "horde: %s[%d]: 499 - sendall(extra header, \\n) failed, %zd\n", hst.name, getpid(), n);
							close(newhandle);
							hfin(EXIT_FAILURE);
							return(EXIT_FAILURE);
						}
					}
				}
				if((n=sendall(newhandle, "\n", 1, 0)))
				{
					if(hst.debug) fprintf(stderr, "horde: %s[%d]: 499 - sendall(head, \\n) failed, %zd\n", hst.name, getpid(), n);
					close(newhandle);
					hfin(EXIT_FAILURE);
					return(EXIT_FAILURE);
				}
				if(h->data) // otherwise assume status does not require one
				{
					n=sendall(newhandle, h->data, h->dlen, 0);
					if(n)
					{
						if(hst.debug) fprintf(stderr, "horde: %s[%d]: 499 - sendall(body) failed, %zd; %s\n", hst.name, getpid(), n, strerror(errno));
						close(newhandle);
						hfin(EXIT_FAILURE);
						return(EXIT_FAILURE);
					}
				}
				hmsg l=new_hmsg("tail", NULL);
				char st[TL_SHORT], sz[TL_SIZET];
				snprintf(st, sizeof(st), "%hu", status);
				add_htag(l, "status", st);
				snprintf(sz, sizeof(sz), "%zu", h->dlen);
				add_htag(l, "bytes", sz);
				if(rpath) add_htag(l, "rpath", rpath);
				if(ip) add_htag(l, "ip", ip);
				add_htag(l, "method", "GET");
				if(ref) add_htag(l, "referrer", ref);
				if(ua) add_htag(l, "user-agent", ua);
				add_htag(l, "date", date);
				gettimeofday(&endtime, NULL);
				double dt=difftime(endtime.tv_sec, starttime.tv_sec)+(endtime.tv_usec-starttime.tv_usec)*1e-6;
				char sdt[16];
				snprintf(sdt, sizeof(sdt), "%.6g", dt);
				add_htag(l, "time", sdt); // this is 'total processing time'
				hsend(1, l);
				free_hmsg(l);
				free_hmsg(h);
				free(rpath);
			break;
			default:
				err(501, "Not Implemented", NULL, newhandle);
				if(hst.debug) fprintf(stderr, "horde: %s[%d]: 501 Not Implemented (%s)\n", hst.name, getpid(), method);
				close(newhandle);
				hfin(EXIT_FAILURE);
				return(EXIT_FAILURE);
			break;
		}
	}
	free(line);
	free(buf);
	if(hst.debug) fprintf(stderr, "horde: %s[%d]: closing conn\n", hst.name, getpid());
	close(newhandle);
	hfin(EXIT_SUCCESS);
	return(EXIT_SUCCESS);
}

void err(unsigned int status, const char *statusmsg, const char *headers, int fd) // should typically only be used if (proc) unavailable
{
	if(!statusmsg)
		statusmsg=http_statusmsg(status);
	if(!statusmsg)
		statusmsg="???";
	char date[256];
	time_t timer = time(NULL);
	struct tm *tm = gmtime(&timer);
	size_t datelen = strftime(date, sizeof(date), "%F %H:%M:%S", tm);
	bool m=true;
	char *buf=malloc(128+datelen+strlen(HTTPD_VERSION)+(strlen(statusmsg))+(headers?strlen(headers):0));
	if(!buf)
	{
		m=false;
		buf="\
HTTP/1.1 500 Internal Server Error\n\
Server: "HTTPD_VERSION" (Unix)\n\
Content-Length: 0\n\
Connection: close\n\
\n";
	}
	else
	{
		sprintf(buf, "\
HTTP/1.1 %d %s\n\
Date: %s\n\
Server: "HTTPD_VERSION" (Unix)\n\
Content-Length: 0\n\
%s\
Connection: close\n\
\n", status, statusmsg, date, headers?headers:"");
	}
	sendall(fd, buf, strlen(buf), 0);
	if(m) free(buf);
}

bool check_msgs(hstate *hst)
{
	fd_set master, readfds;
	FD_ZERO(&master);
	FD_SET(STDIN_FILENO, &master);
	struct timeval timeout;
	timeout.tv_sec=0;
	timeout.tv_usec=0;
	readfds=master;
	select(STDIN_FILENO+1, &readfds, NULL, NULL, &timeout);
	if(FD_ISSET(STDIN_FILENO, &readfds))
	{
		char *buf=getl(STDIN_FILENO);
		if(!buf) return(false);
		if(*buf)
		{
			hmsg h=hmsg_from_str(buf, true);
			if(h)
			{
				if(!hmsg_state(h, hst))
					fprintf(stderr, "horde: %s[%d]: Unrecognised funct %s in check_msgs()\n", hst->name, getpid(), h->funct);
				free_hmsg(h);
			}
			else
			{
				if(hst->debug) fprintf(stderr, "horde: %s[%d]: Nonsense hmsg received\n", hst->name, getpid());
				if(hst->debug) fprintf(stderr, "\t%s\n", buf);
			}
		}
		free(buf);
		return(true);
	}
	return(false);
}

ssize_t sendall(int sockfd, const void *buf, size_t length, int flags)
{
	const char *p=buf;
	ssize_t left=length;
	ssize_t n;
	while((n=send(sockfd, p, left, flags))<left)
	{
		if(n<0)
			return(n);
		p+=n;
		left-=n;
	}
	return(0);
}
